Remove hacky_wrapper from VariableType and TraceType

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44158 VariableKernel calls into scattered C++ api
* #44087 Add faithful C++ API
* #44071 Trace scattered tensor options arguments
* #44062 Remove hacky_wrapper from BackendSelect kernels
* **#44005 Remove hacky_wrapper from VariableType and TraceType**

Previously, VariableType and TraceType kernels were still written in the legacy way, i.e. they took one TensorOptions argument instead of scattered dtype, layout, device, pin_memory,  and they used hacky_wrapper to be callable.

Now with this PR, variable and tracing kernels are written in the new way and no hacky_wrapper is needed for them.

This only affects ops with `use_c10_dispatcher: full`.

Differential Revision: [D23466042](https://our.internmc.facebook.com/intern/diff/D23466042/)

smessmer committed:
Remove hacky_wrapper from VariableType and TraceType

Previously, VariableType and TraceType kernels were still written in the legacy way, i.e. they took one TensorOptions argument instead of scattered dtype, layout, device, pin_memory,  and they used hacky_wrapper to be callable. This caused a re-wrapping step. Calling into a Variable or Tracing kernel required taking the individual scattered arguments, packing them into a TensorOptions, and the kernel itself then likely re-dispatched, scattering those again into individual arguments.

Now with this PR, variable and tracing kernels are written in the new way and no hacky_wrapper or rewrapping is needed for them.

Differential Revision: [D23466042](https://our.internmc.facebook.com/intern/diff/D23466042/)

[ghstack-poisoned]

smessmer committed:
Update on "Remove hacky_wrapper from VariableType and TraceType"


Previously, VariableType and TraceType kernels were still written in the legacy way, i.e. they took one TensorOptions argument instead of scattered dtype, layout, device, pin_memory,  and they used hacky_wrapper to be callable.

Now with this PR, variable and tracing kernels are written in the new way and no hacky_wrapper is needed for them.

Differential Revision: [D23466042](https://our.internmc.facebook.com/intern/diff/D23466042/)

[ghstack-poisoned]

bhosmer reviewed (approved):
A few minor things inline, plus one thing about tracing TensorOptions (I think this relates to a semantics issue you mentioned in chat, but I thought the answer was that we could scatter them into addInput()).

Also I'd add to the PR description that the modification is specifically for use_c10_dispatcher: full.

Note: I think both build failures come down to
```
Sep 02 19:19:41 -- Build files have been written to: /var/lib/jenkins/workspace/build_test_custom_build/predictor
Sep 02 19:19:41 + make
Sep 02 19:19:41 Scanning dependencies of target Predictor
Sep 02 19:19:41 [ 50%] Building CXX object CMakeFiles/Predictor.dir/predictor.cpp.o
Sep 02 19:19:41 [100%] Linking CXX executable Predictor
Sep 02 19:19:44 [100%] Built target Predictor
Sep 02 19:19:44 + run_predictor
Sep 02 19:19:44 + cd /var/lib/jenkins/workspace/build_test_custom_build/predictor
Sep 02 19:19:44 + ./Predictor /var/lib/jenkins/workspace/build_test_custom_build/MobileNetV2.pt
Sep 02 19:19:44 terminate called after throwing an instance of 'torch::jit::ErrorReport'
Sep 02 19:19:44   what():  
Sep 02 19:19:44 
Sep 02 19:19:44 aten::_convolution(Tensor input, Tensor weight, Tensor? bias, int[] stride, int[] padding, int[] dilation, bool transposed, int[] output_padding, int groups, bool benchmark, bool deterministic, bool cudnn_enabled) -> (Tensor):
Sep 02 19:19:44 Expected at most 12 arguments but found 13 positional arguments.
```

ezyang commented on tools/autograd/gen_variable_type.py:
```diff
@@ -491,6 +494,11 @@ def format_trace_op_name(declaration):
+            if value == "options":
+                value = "TensorOptions().dtype(dtype).layout(layout).device(device).pinned_memory(pin_memory)"
+        else:
+            assert declaration['use_c10_dispatcher'] == 'with_codegenerated_unboxing_wrapper'
```
Not in `tools/autograd` yet, alas.

bhosmer replied:
RIght, it's still raw data over here. But since we're now generating Declarations.yaml from groomed data, it seems reasonable to yank some of these (effectively no-op) sanity checks when it really helps to declutter the code. 

Not a huge deal either way, fine by me if you'd rather have them as an inducement to bring the data model over :)

ezyang replied:
IDK, the assert is not making a statement about how clean or dirty the upstream source of the data is; it is there to make sure that *you* (the consumer) and upstream (the producer) agree on what the invariants are supposed to be, and complain loudly when they mismatch.

smessmer commented on tools/autograd/gen_variable_type.py:
```diff
@@ -679,12 +693,17 @@ def gen_variable_type_shard(out, aten_declarations, template_path, suffix, heade
+        if declaration['use_c10_dispatcher'] == 'full':
+            formals = declaration['schema_order_formals']
+        else:
+            assert declaration['use_c10_dispatcher'] == 'with_codegenerated_unboxing_wrapper'
```
IMHO, an assertion is always strictly better than a comment since comments age and become inaccurate

smessmer commented on tools/autograd/gen_variable_type.py:
```diff
@@ -1222,7 +1241,11 @@ def requires_unpack(arg):
+            if declaration['use_c10_dispatcher'] == 'full':
+                body.append(UNPACK_OPTIONS.substitute(arg_name=arg['name']))
+            else:
+                declaration['use_c10_dispatcher'] == 'with_codegenerated_unboxing_wrapper'
```
ugh seems I forgot to type the `assert` here. fixing.

smessmer committed:
Update on "Remove hacky_wrapper from VariableType and TraceType"


Previously, VariableType and TraceType kernels were still written in the legacy way, i.e. they took one TensorOptions argument instead of scattered dtype, layout, device, pin_memory,  and they used hacky_wrapper to be callable.

Now with this PR, variable and tracing kernels are written in the new way and no hacky_wrapper is needed for them.

This only affects ops with `use_c10_dispatcher: full`.

Differential Revision: [D23466042](https://our.internmc.facebook.com/intern/diff/D23466042/)

[ghstack-poisoned]

smessmer committed:
Update on "Remove hacky_wrapper from VariableType and TraceType"


Previously, VariableType and TraceType kernels were still written in the legacy way, i.e. they took one TensorOptions argument instead of scattered dtype, layout, device, pin_memory,  and they used hacky_wrapper to be callable.

Now with this PR, variable and tracing kernels are written in the new way and no hacky_wrapper is needed for them.

This only affects ops with `use_c10_dispatcher: full`.

Differential Revision: [D23466042](https://our.internmc.facebook.com/intern/diff/D23466042/)

[ghstack-poisoned]

smessmer committed:
Update on "Remove hacky_wrapper from VariableType and TraceType"


Previously, VariableType and TraceType kernels were still written in the legacy way, i.e. they took one TensorOptions argument instead of scattered dtype, layout, device, pin_memory,  and they used hacky_wrapper to be callable.

Now with this PR, variable and tracing kernels are written in the new way and no hacky_wrapper is needed for them.

This only affects ops with `use_c10_dispatcher: full`.

Differential Revision: [D23466042](https://our.internmc.facebook.com/intern/diff/D23466042/)

[ghstack-poisoned]
